File: 347.top-k-frequent-elements.py
# @leet start
import heapq
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        elems = {}
        for i in nums:
            if i not in elems:
                elems[i] = 0
            elems[i] += 1

        vals = [(-v, k) for k,v in zip(elems.keys(), elems.values())]
        heapq.heapify(vals)

        result = []
        for i in range(k):
            result.append(heapq.heappop(vals)[1])

        return result

# @leet end

File: 347.top-k-frequent-elements.cpp
// @leet start
#include <unordered_map>

bool comp(pair<int, int> a, pair<int,int> b) {
    return a.second < b.second;
}
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        std::unordered_map<int, int> map;

        for (int num : nums) {
            map[num]++;
        }

        vector<pair<int, int>> list;
        for (const auto &p : map) {
            list.push_back(p);
        }

        make_heap(list.begin(), list.end(), comp);
        vector<int> result;

        for (int i = 0; i < k; i++) {
            pop_heap(list.begin(), list.end(), comp);
            result.push_back(list.back().first);
            list.pop_back();
        }

        return result;
    }
};
// @leet end
